File: packages/modules/ImsMedia/framework/src/android/telephony/imsmedia/RtpContextParams.java
/**
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.telephony.imsmedia;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Objects;

/**
 * Class to encapsulate RTP (Real-time Transport Protocol) parameters
 * required for RTP context transfer to maintain RTP stream continuity
 *
 * @hide
 */

public final class RtpContextParams implements Parcelable {
    /** Synchronization source (SSRC) identifier of RTP stream **/
    private final long mSsrc;
    /** The timestamp reflects the sampling instant of the last RTP packet sent **/
    private final long mTimestamp;
    /** The sequence number of last RTP packet sent **/
    private final int mSequenceNumber;
    /** Maximum 32bit value **/
    private static final long MAX_VALUE_32BIT = 4294967295L;
    /** Maximum 16bit value **/
    private static final int MAX_VALUE_16BIT = 65535;

    /** @hide **/
    public RtpContextParams(Parcel in) {
        mSsrc = in.readLong();
        mTimestamp = in.readLong();
        mSequenceNumber = in.readInt();
    }

    private RtpContextParams(long ssrc, long timestamp, int sequenceNumber) {
        this.mSsrc = ssrc;
        this.mTimestamp  = timestamp;
        this.mSequenceNumber = sequenceNumber;
    }

    /** @hide **/
    public long getSsrc() {
        return mSsrc;
    }

    /** @hide **/
    public long getTimestamp() {
        return mTimestamp;
    }

    /** @hide **/
    public int getSequenceNumber() {
        return mSequenceNumber;
    }

    @NonNull
    @Override
    public String toString() {
        return "RtpContextParams: {mSsrc=" + mSsrc
            + ", mTimestamp=" + mTimestamp
            + ", mSequenceNumber=" + mSequenceNumber
            + " }";
    }

    @Override
    public int hashCode() {
        return Objects.hash(mSsrc, mTimestamp, mSequenceNumber);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (o == null || !(o instanceof RtpContextParams) || hashCode() != o.hashCode()) {
            return false;
        }

        if (this == o) {
            return true;
        }

        RtpContextParams s = (RtpContextParams) o;

        return (mSsrc == s.mSsrc
            && mTimestamp == s.mTimestamp
            && mSequenceNumber == s.mSequenceNumber);
    }

    /**
     * {@link Parcelable#describeContents}
     */
    public int describeContents() {
        return 0;
    }

    /**
     * {@link Parcelable#writeToParcel}
     */
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(mSsrc);
        dest.writeLong(mTimestamp);
        dest.writeInt(mSequenceNumber);
    }

    public static final @NonNull Parcelable.Creator<RtpContextParams>
            CREATOR = new Parcelable.Creator() {
                public RtpContextParams createFromParcel(Parcel in) {
                    // TODO use builder class so it will validate
                    return new RtpContextParams(in);
                }

                public RtpContextParams[] newArray(int size) {
                    return new RtpContextParams[size];
                }
            };

    /**
     * Provides a convenient way to set the fields of a {@link RtpContextParams}
     * when creating a new instance.
     */
    public static final class Builder {
        private long mSsrc;
        private long mTimestamp;
        private int mSequenceNumber;

        /**
         * Default constructor for Builder.
         */
        public Builder() {
        }

        /**
         * Set the Synchronization source (SSRC) identifier of RTP stream
         * Throws IllegalArgumentException if SSRC is set greater than maximum 32bit value
         *
         * @param ssrc ssrc of RTP stream
         * @return The same instance of the builder.
         */
        public @NonNull Builder setSsrc(final long ssrc) {
            if (ssrc < 0 || ssrc > MAX_VALUE_32BIT) {
                throw new IllegalArgumentException("Invalid ssrc value: " + ssrc);
            } else {
                this.mSsrc = ssrc;
            }
            return this;
        }

        /**
         * Set the timestamp
         * Throws IllegalArgumentException if timestamp is set greater than maximum 32bit value
         *
         * @param timestamp The timestamp of last RTP packet sent
         * @return The same instance of the builder.
         */
        public @NonNull Builder setTimestamp(final long timestamp) {
            if (timestamp < 0 || timestamp > MAX_VALUE_32BIT) {
                throw new IllegalArgumentException("Invalid timestamp value: " + timestamp);
            } else {
                this.mTimestamp = timestamp;
            }
            return this;
        }

        /**
         * Set the Sequence Number
         * Throws IllegalArgumentException if sequenceNumber is set greater than maximum 16bit value
         *
         * @param sequenceNumber The sequence number of last RTP packet sent
         * @return The same instance of the builder.
         */
        public @NonNull Builder setSequenceNumber(final int sequenceNumber) {
            if (sequenceNumber < 0 || sequenceNumber > MAX_VALUE_16BIT) {
                throw new IllegalArgumentException(
                        "Invalid sequenceNumber value: " + sequenceNumber);
            } else {
                this.mSequenceNumber = sequenceNumber;
            }
            return this;
        }

        /**
         * Build the RtpContextParams.
         *
         * @return the RtpContextParams object.
         */
        public @NonNull RtpContextParams build() {
            // TODO validation
            return new RtpContextParams(mSsrc, mTimestamp, mSequenceNumber);
        }
    }
}

File: packages/modules/ImsMedia/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/src/RtpContextParams.cpp

#include <RtpContextParams.h>

namespace android
{

namespace telephony
{

namespace imsmedia
{

RtpContextParams::RtpContextParams() :
        ssrc(0),
        timestamp(0),
        sequenceNumber(0)
{
}

RtpContextParams::RtpContextParams(
        const int64_t ssrc, const int64_t timestamp, const int32_t sequenceNumber)
{
    this->ssrc = ssrc;
    this->timestamp = timestamp;
    this->sequenceNumber = sequenceNumber;
}

RtpContextParams::RtpContextParams(const RtpContextParams& params)
{
    this->ssrc = params.ssrc;
    this->timestamp = params.timestamp;
    this->sequenceNumber = params.sequenceNumber;
}

RtpContextParams::~RtpContextParams() {}

RtpContextParams& RtpContextParams::operator=(const RtpContextParams& params)
{
    if (this != &params)
    {
        this->ssrc = params.ssrc;
        this->timestamp = params.timestamp;
        this->sequenceNumber = params.sequenceNumber;
    }
    return *this;
}

bool RtpContextParams::operator==(const RtpContextParams& params) const
{
    return (this->ssrc == params.ssrc && this->timestamp == params.timestamp &&
            this->sequenceNumber == params.sequenceNumber);
}

bool RtpContextParams::operator!=(const RtpContextParams& params) const
{
    return (this->ssrc != params.ssrc || this->timestamp != params.timestamp ||
            this->sequenceNumber != params.sequenceNumber);
}

status_t RtpContextParams::writeToParcel(Parcel* out) const
{
    status_t err;

    if (out == nullptr)
    {
        return BAD_VALUE;
    }

    err = out->writeInt64(ssrc);

    if (err != NO_ERROR)
    {
        return err;
    }

    err = out->writeInt64(timestamp);

    if (err != NO_ERROR)
    {
        return err;
    }

    err = out->writeInt32(sequenceNumber);

    if (err != NO_ERROR)
    {
        return err;
    }

    return NO_ERROR;
}

status_t RtpContextParams::readFromParcel(const Parcel* in)
{
    status_t err;

    if (in == nullptr)
    {
        return BAD_VALUE;
    }

    err = in->readInt64(&ssrc);

    if (err != NO_ERROR)
    {
        return err;
    }

    err = in->readInt64(&timestamp);

    if (err != NO_ERROR)
    {
        return err;
    }

    err = in->readInt32(&sequenceNumber);

    if (err != NO_ERROR)
    {
        return err;
    }

    return NO_ERROR;
}

int64_t RtpContextParams::getSsrc()
{
    return ssrc;
}

void RtpContextParams::setSsrc(int64_t ssrc)
{
    this->ssrc = ssrc;
}

int64_t RtpContextParams::getTimestamp()
{
    return timestamp;
}

void RtpContextParams::setTimestamp(int64_t timestamp)
{
    this->timestamp = timestamp;
}

int32_t RtpContextParams::getSequenceNumber()
{
    return sequenceNumber;
}

void RtpContextParams::setSequenceNumber(int32_t sequenceNumber)
{
    this->sequenceNumber = sequenceNumber;
}

void RtpContextParams::setDefaultConfig()
{
    ssrc = 0;
    timestamp = 0;
    sequenceNumber = 0;
}

}  // namespace imsmedia

}  // namespace telephony

}  // namespace android